ratatui-widgets/src/paragraph.rs

```diff
@@ -489,6 +497,35 @@ impl<'a> Paragraph<'a> {
     }
 }
 
+fn render_text<'a, C: LineComposer<'a>>(mut composer: C, area: Rect, buf: &mut Buffer) {
+    let mut y = 0;
+    while let Some(WrappedLine {
+        line: current_line,
+        width: current_line_width,
+        alignment: current_line_alignment,
+    }) = composer.next_line()
+    {
+        let mut x = get_line_offset(current_line_width, area.width, current_line_alignment);
+        for StyledGrapheme { symbol, style } in current_line {
+            let width = symbol.width();
+            if width == 0 {
+                continue;
+            }
+            // If the symbol is empty, the last char which rendered last time will
+            // leave on the line. It's a quick fix.
+            let symbol = if symbol.is_empty() { " " } else { symbol };
+            buf[(area.left() + x, area.top() + y)]
+                .set_symbol(symbol)
+                .set_style(*style);
+            x += width as u16;
+        }
```

Comment:
This chunk seems like an exact duplicate. Is there a way to not duplicate this (extract a method). (Side note: In general, we arrange supporting child methods below the caller)